Create documentation for defaults.yaml (#146)
bakdata/kpops: docs/docs/user/references/defaults.md

# Defaults

KPOps has a very efficient way of dealing with repeating settings which manifests as [`defaults.yaml`](/resources/pipeline-defaults/defaults){target=_blank}. This file provides the user with the power to set defaults for any and all [components](/user/references/components){target=_blank}, thus omitting the need to repeat the same settings in [`pipeline.yaml`](/resources/pipeline-components/pipeline){target=_blank}.

[See real-world examples for `defaults`](/resources/examples/defaults){target=_blank}.

??? "KPOps component Hierarchy"
    --8<--
    ./docs/resources/architecture/components-hierarchy.md

>     --8<--

Reviewer: Do we want the figure in the info box?

Author: I prefer it like that. "Defaults "is not the main page that discusses components, so their relationships are better explained in "Components" where the pic is not collapsed. 

imo it's nice to have the figure in "Defaults" for a quick look, but it is redundant to have it full-sized by default.

## Features

### Inheritance

An important mechanic of KPOps is that `defaults` set for a component apply to all components that inherit from it.

It is possible, although not recommended, to add settings that are specific to a component's subclass. An example would be configuring `offsetTopic` under `kafka-connect` instead of `kafka-source-connector`.

### Configuration

> !!! tip inline end

Author: fixed

!!! tip
    `defaults` is the default value of `defaults_filename_prefix`.
    Together with `defaults_path` and  `environment` it can be changed in [`config.yaml`](/user/references/config/#__codelineno-0-16){target=_blank}

## Components

<!-- When possible, automatically generate a list of all component-specific settings under each component. -->

The `defaults` codeblocks in this section contain the full set of settings that are specific to the component. If a setting already exists in a parent config, it will not be included in the child's.

### [KubernetesApp](/user/references/components/#kubernetesapp){target=_blank}

??? "`defaults.yaml`"

    ```yaml
      --8<--
      ./docs/resources/pipeline-defaults/defaults-kubernetes-app.yaml
      --8<--
    ```

### [KafkaApp](/user/references/components/#kafkaapp){target=_blank}

??? "`defaults.yaml`"

    ```yaml
      --8<--
      ./docs/resources/pipeline-defaults/defaults-kafka-app.yaml
      --8<--
    ```

### [StreamsApp](/user/references/components/#streamsapp){target=_blank}

> ??? "`defaults.yaml`"

Reviewer: Instead of a note box you can use a cool example box: https://squidfunk.github.io/mkdocs-material/reference/admonitions/#type:example
WDYT?

Author: Makes sense, will do

Author: done

    ```yaml
      --8<--
      ./docs/resources/pipeline-defaults/defaults-streams-app.yaml
      --8<--
    ```

### [ProducerApp](/user/references/components/#producerapp)

??? "`defaults.yaml`"

    ```yaml
      --8<--
      ./docs/resources/pipeline-defaults/defaults-producer.yaml
      --8<--
    ```

> ### [KafkaConnector](#kafkaconnector)

Author: But then all the others are blue and this one isn't, it makes reading the doc a bit harder IMO as you expect the section headings to all be the same

> ### [KafkaConnector](#kafkaconnector)

Reviewer 2: hm, but might be more confusing to users if there is a link that doesn't do anything

> ### [KafkaConnector](#kafkaconnector)

Reviewer: I agree with salomon

> ### [KafkaConnector](#kafkaconnector)

Author: Removed

# Defaults

KPOps has a very efficient way of dealing with repeating settings which manifests as [`defaults.yaml`](/resources/pipeline-defaults/defaults).

> This file provides the user with the power to set defaults for any and all components, thus omitting the need to repeat the same settings in [`pipeline.yaml`](/user/references/components).

Reviewer 2: you're off to a good start here!

please also go into detail about the inheritance model, i.e. configuring a base component as shared config of its subcomponents.

You can use an example to further clarify this, e.g. source & sink connector with their base class `kafka-connect` containing the shared config for both

Author: Do you think the current examples are sufficient?

## [KubernetesApp](/user/references/components/#kubernetesapp)

??? "`defaults.yaml`"

    ```yaml
      --8<--
      ./docs/resources/pipeline-defaults/defaults-kubernetes-app.yaml
      --8<--
    ```

## [KafkaApp](/user/references/components/#kafkaapp)

??? "`defaults.yaml`"

    ```yaml
      --8<--
      ./docs/resources/pipeline-defaults/defaults-kafka-app.yaml
      --8<--
    ```

## [StreamsApp](/user/references/components/#streamsapp)

??? "`defaults.yaml`"

    ```yaml
      --8<--
      ./docs/resources/pipeline-defaults/defaults-streams-app.yaml
      --8<--
    ```

## [ProducerApp](/user/references/components/#producerapp)

??? "`defaults.yaml`"

    ```yaml
      --8<--
      ./docs/resources/pipeline-defaults/defaults-producer.yaml
      --8<--
    ```

## [KafkaConnector](#kafkaconnector)

`KafkaConnector` is a component that deploys [Kafka Connectors](https://kafka.apache.org/documentation.html#connect_configuring). Since a connector cannot be different from [sink](#kafkasinkconnector) or [source](#kafkasourceconnector), it is not recommended to use `KafkaConnector` for deployment in [`pipeline.yaml`](/resources/pipeline-components/pipeline).

> Instead, `KafkaConnector` should be used in [`defaults.yaml`](/resources/pipeline-defaults/defaults) to set defaults for all connectors in the pipeline as they usually share a lot common settings.

Reviewer 2: ```suggestion
Instead, `KafkaConnector` should be used in [`defaults.yaml`](/resources/pipeline-defaults/defaults) to set defaults for all connectors in the pipeline as they can share some common settings.
```

# Defaults

> KPOps has a very efficient way of dealing with repeating settings which manifests as [`defaults.yaml`](/resources/pipeline-defaults/defaults). This file provides the user with the power to set defaults for any and all [components](/user/references/components), thus omitting the need to repeat the same settings in [`pipeline.yaml`](/user/references/components).

Reviewer 2: would it make sense to also link to `/resources/pipeline-components/pipeline` as `pipeline.yaml` reference?

??? "KPOps component Hierarchy"
    --8<--
    ./docs/resources/architecture/components-hierarchy.md
    --8<--

> A very important mechanic of KPOps' defaults is that defaults set for a component apply to all components that inherit from it. An example can be found [here](#examples).

Reviewer 2: consider moving the example up so that it sits right after this paragraph. jumping around a lot could make it difficult

Author: WDYT?

Reviewer 2: looks good to me

KPOps has a very efficient way of dealing with repeating settings which manifests as [`defaults.yaml`](/resources/pipeline-defaults/defaults). This file provides the user with the power to set defaults for any and all [components](/user/references/components), thus omitting the need to repeat the same settings in [`pipeline.yaml`](/user/references/components).

A very important mechanic of KPOps' defaults is that defaults set for a component apply to all components that inherit from it. An example can be found [here](https://github.com/bakdata/kpops-examples/tree/main/word-count/deployment/kpops/defaults).

??? "KPOps component Hierarchy"
    --8<--
    ./docs/resources/architecture/components-hierarchy.md
    --8<--

!!! tip inline end
    `defaults` is the default value of `defaults_filename_prefix`.

>     It, together with `defaults_path` and  `environment` can be changed in [`config.yaml`](/user/references/config)

Reviewer 2: ```suggestion
    Together with `defaults_path` and `environment` it can be changed in [`config.yaml`](/user/references/config).
```
reads a little easier :)

KPOps has a very efficient way of dealing with repeating settings which manifests as [`defaults.yaml`](/resources/pipeline-defaults/defaults). This file provides the user with the power to set defaults for any and all [components](/user/references/components), thus omitting the need to repeat the same settings in [`pipeline.yaml`](/resources/pipeline-components/pipeline).

> See real-world `defaults` configs [here](/resources/examples/defaults).

Reviewer 2: how about 'see real-word examples for `defaults`'?

> See real-world `defaults` configs [here](/resources/examples/defaults).

Author: Sounds good
